### PR TITLE
Counting Islands Assignment 4 - Draft

### PR DESCRIPTION
**Pull request by Veleslavia**
Counting islands by recursively visiting connected land areas. 
That recursive algorithm looping over the sea matrix and if we have a land, we should process the corresponding island and add +1 to the total number of islands.

The optional part is not yet implemented.

### CODE REVIEW

File: veleslavia/assignment-4/arbitrary_seamap.py [hunk 1]
```diff
@@ -0,0 +1,3 @@
+class ArbitrarySeaMap:
```
Comment (jdb) [inline veleslavia/assignment-4/arbitrary_seamap.py]: Do you really want to submit this file? It looks empty.

File: veleslavia/assignment-4/seamap.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
+from typing import List
+import unittest
+
+
+def process_island(sea_map: List[List[bool]], i: int, j: int):
+    connected_directions = [(0, -1), (0, 1), (1, 0), (-1, 0)]
+    sea_map[i][j] = False
```
Comment (jdb) [inline veleslavia/assignment-4/seamap.py]: Oh ! pretty surprising here: the input is mutated :)
If the caller gives this function a reference to the map, then after the functions runs, the map is full sea !
- Either you need to keep another data structure to remember which area you have already visited,
- or you warn the user of the function that the user needs to keep a copy of the map before sending it to your function.

File: veleslavia/assignment-4/seamap.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    for (move_i, move_j) in connected_directions:
+        if 0 <= i+move_i < len(sea_map) and 0 <= j+move_j < len(sea_map[i+move_i]):
+            if sea_map[i+move_i][j+move_j]:
+                process_island(sea_map, i+move_i, j+move_j)
+
+
+def count_islands(sea_map: List[List[bool]]) -> int:
```
Comment (jdb) [inline veleslavia/assignment-4/seamap.py]: Please add docstrings !

File: veleslavia/assignment-4/seamap.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    n_islands = 0
+    for i in range(len(sea_map)):
+        for j in range(len(sea_map[i])):
+            # if land
```
Comment (jdb) [inline veleslavia/assignment-4/seamap.py]: The comment is misleading: either this is land, OR this was already visited!

File: veleslavia/assignment-4/seamap.py (hunk 3 above)
Comment (Veleslavia) [reply veleslavia/assignment-4/seamap.py]: Done

File: veleslavia/assignment-4/seamap.py (hunk 4 above)
Comment (Veleslavia) [reply veleslavia/assignment-4/seamap.py]: Added another structure, sea_map mutation is misleading indeed

File: veleslavia/assignment-4/seamap.py (hunk 3 above)
Comment (jdb) [reply veleslavia/assignment-4/seamap.py]: Well, you should add docstrings to all public functions in the module. Ice on the cake if you add also a module docstring  !

File: veleslavia/assignment-4/seamap.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
+from itertools import product
+from typing import List
+import unittest
+
```
Comment (jdb) [inline veleslavia/assignment-4/seamap.py]: Try to present a short description of your approach, say if you are in an interview, then you really want to nail the communication with the interviewer. This is one part of the interview that is arguably as important as writing a solution. Among the question we are requested to ask ourselves as interviewer is: would I want to work with this person? would I work well with this person?

File: veleslavia/assignment-4/seamap.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+def count_islands(sea_map: List[List[bool]]) -> int:
+    """
+    Count number of islands in the map.
+    :param sea_map: 2D map of booleans where True represents LAND and False represents SEA
+    :return: number of islands in the map
+    """
+    visited = [[False] * len(row) for row in sea_map]
+    return sum(process_island(sea_map, visited, i, j) for (i, j) in
```
Comment (jdb) [inline veleslavia/assignment-4/seamap.py]: nit: I suggest to write each element of the generator expression on each line:

return sum(
    process_island(sea_map, visited, i, j)
    for (i, j) in product(range(len(sea_map)), range(len(sea_map[0])))
    if sea_map[i][j] and not visited[i][j])

The generator expression is 3 elements:
- the expression that executed at each loop
- the for part
- the if part

Comment (Veleslavia) [reply veleslavia/assignment-4/seamap.py]: Updated with all your comments :)

**Pull request merged**